### PR TITLE
Don't omit `Content-Length` header for `Content-Length: 0` cases

### LINKED ISSUE #1099: Missing content-length header field in some responses
### Checklist

- [x] The bug is reproducible against the latest release and/or `master`.
- [x] There are no similar issues or pull requests to fix it yet.

### Describe the bug

The Content-Length header field is not set for cases where body is empty. In some cases this seems to not be according to the HTTP/1.1 spec and creates troubles when using for example traefik as a proxy in front of a starlette-based backend.

See this issue for details: https://github.com/traefik/traefik/issues/4456
Specifically, the final conclusion: https://github.com/traefik/traefik/issues/4456#issuecomment-593539169

### To reproduce

```python
>>> from starlette.responses import RedirectResponse
>>> RedirectResponse(url="/", status_code=302).headers
MutableHeaders({'location': '/'})
```
Header has not content-length field.

### Expected behavior

```python
>>> from starlette.responses import RedirectResponse
>>> RedirectResponse(url="/", status_code=302).headers
MutableHeaders({'content-length': '0', 'location': '/'})
```
I expect there to be a content-length field specifying zero length

### Actual behavior

```python
>>> from starlette.responses import RedirectResponse
>>> RedirectResponse(url="/", status_code=302).headers
MutableHeaders({'location': '/'})
```
Header has no content-length field.

### Debugging material

N/A

### Environment

- OS: Linux (WSL2)
- Python version: 3.8.2
- Starlette version: 0.13.6

### Additional context

This bug is probably also present for other type of responses (other status-codes). Last paragraph (in bold) from https://tools.ietf.org/html/rfc7230#section-3.3 is of interest:

> 3.3.  Message Body
> 
>    The message body (if any) of an HTTP message is used to carry the
>    payload body of that request or response.  The message body is
>    identical to the payload body unless a transfer coding has been
>    applied, as described in Section 3.3.1.
> 
>      message-body = *OCTET
> 
>    The rules for when a message body is allowed in a message differ for
>    requests and responses.
> 
>    The presence of a message body in a request is signaled by a
>    Content-Length or Transfer-Encoding header field.  Request message
>    framing is independent of method semantics, even if the method does
>    not define any use for a message body.
> 
>    The presence of a message body in a response depends on both the
>    request method to which it is responding and the response status code
>    (Section 3.1.2).  Responses to the HEAD request method (Section 4.3.2
>    of [RFC7231]) never include a message body because the associated
>    response header fields (e.g., Transfer-Encoding, Content-Length,
>    etc.), if present, indicate only what their values would have been if
>    the request method had been GET (Section 4.3.1 of [RFC7231]). 2xx
>    (Successful) responses to a CONNECT request method (Section 4.3.6 of
>    [RFC7231]) switch to tunnel mode instead of having a message body.
>    **All 1xx (Informational), 204 (No Content), and 304 (Not Modified)
>    responses do not include a message body.  All other responses do
>    include a message body, although the body might be of zero length.**

### PR DESCRIPTION
**Pull request by Kludex**
This PR adds the `content-length` header by default, if not already included. For more information, please check on #1099 and the referred RFC.

Next step would be to add a conditional to check if `status_code` < 200 and `status_code` != (204, 304)`

Closes #1099

### CODE REVIEW

File: starlette/responses.py [hunk 1]
```diff
@@ -289,7 +289,7 @@ def set_stat_headers(self, stat_result: os.stat_result) -> None:
         etag_base = str(stat_result.st_mtime) + "-" + str(stat_result.st_size)
         etag = hashlib.md5(etag_base.encode()).hexdigest()
 
-        self.headers.setdefault("content-length", content_length)
+        self.headers["content-length"] = content_length
```
Comment (Kludex) [inline starlette/responses.py]: We are sure that `content-length` exists at this point, considering the change of conditional in the `init_headers()`.

**Comment by lovelydinosaur** [comment]
Okay, so forgetting for a moment about `1xx`, `204`, `304`, here's a sensible set of test cases that I can think of wrt. setting `Content-Length`...

* `Response()` - Regular empty response. `Content-Length: 0`  (`RedirectResponse` is a good real-world example of this case)
* `Response(content='...')` - Regular non-empty response. `Content-Length: x`
* `FileResponse(...)`  - File response, unknown length. `Transfer-Encoding: chunked`.
* `FileResponse(..., stat_result=...)`  - File response, known length. `Content-Length: x`
* `StreamingResponse(...)` - Streaming response, unknown length. `Transfer-Encoding: chunked`
* `StreamingResponse(..., headers={"Content-Length": "123"})` - Streaming response, known length. `Content-Length: x`

For `Transfer-Encoding: chunked` cases it's also potentially okay to just omit the header, since I *think* the ASGI server is responsible for adding that, but I've not fully verified that.

I'd suggest the next step here is expanding the tests in this PR so that we're covering each case, and can see which ones need fixing up.

File: starlette/responses.py (hunk 1 above)
Comment (lovelydinosaur) [reply starlette/responses.py]: I'd be a bit careful about that assumption. Personally I'd *start* by reverting this bit, making sure we've got the different test cases I've described. The `StreamingResponse` and `FileResponse` won't yet pass, but let's look at *how* we should resolve them once we've identified the test failures fully.

File: tests/test_responses.py [hunk 2]
```diff
@@ -309,3 +324,8 @@ def test_head_method(test_client_factory):
     client = test_client_factory(app)
     response = client.head("/")
     assert response.text == ""
+
+
+def test_empty_response():
```
Comment (lovelydinosaur) [inline tests/test_responses.py]: This is the test case that we should probably expand to cover the 6 different cases described. I guess we want these test cases?...

* `test_empty_response`
* `test_non_empty_response`
* `test_file_response_unknown_size`
* `test_file_response_known_size`
* `test_streaming_response_unknown_size`
* `test_streaming_response_known_size`

**Comment by lovelydinosaur** [comment]
I've retitled this PR to be more representative of the change.

File: tests/test_responses.py (hunk 2 above)
Comment (Kludex) [reply tests/test_responses.py]: When do we have a `FileResponse` with unknown size?

**Comment by Kludex** [comment]
Expressing the thoughts I've shared on Gitter here as well:

I need to be sure that the response class is not `StreamingResponse` or `FileResponse` before adding the `Content-Length: 0`, because those four cases are already being handled now (file response unknown size, file response known size, streaming response known size, and streaming response unknown size). 

I don't have (or I can't see) any attribute I can check to be sure it's not a "Transfer-Encoding: chunked" on the at the `init_headers()` point.

I've added a conditional on the `Response.init_headers()` to check if the response class was one of those two before adding the `content-length: 0` headers... Can we do better?

**Comment by Kludex** [comment]
Updated the solution to check if body is present on the response. If it's not, then it doesn't add the `Content-Lenght: 0` headers.

**Review by lovelydinosaur — APPROVED** [review]
Great! This is a really nicely targeted fix now. 👍

Possibly improvements still for the 1xx, 204, 304 cases, but we can consider those as a separate conversation.

**Comment by Kludex** [comment]
For future reference, some discussions about this PR can be retrieved on Gitter: https://gitter.im/encode/community?at=61d6e204526fb77b3166a445

**Comment by Kludex** [comment]
I've removed the period at the end to squash the commit.

Thanks @tomchristie :)

File: tests/test_responses.py [hunk 3]
```diff
@@ -309,3 +324,45 @@ def test_head_method(test_client_factory):
...
+
+
+def test_non_empty_response(test_client_factory):
+    app = Response(content="hi")
+    client: TestClient = test_client_factory(app)
+    response = client.get("/")
+    assert response.headers["content-length"] == "2"
+
+
+def test_file_response_known_size(tmpdir, test_client_factory):
```
Comment (lovelydinosaur) [inline tests/test_responses.py:343]: I didn't realise at the time, but we could just name this one `test_file_response`.

Comment (Kludex) [reply tests/test_responses.py:343]: There's one above already (that also checks the `content-length` being present), maybe it's just redundant.

**Pull request merged**